### PR TITLE
Add mechanism to forward fastrtps logging into rcl logging system.

### PR DESCRIPTION
**Pull request by windelbouwman**
This change forwards the fastrtps logging into the rcl logging. Note that you require a debug build of fastrtps to receive logging of detailed levels.

This is more a proof-of-concept. I'm not sure what the appropriate way is.

### CODE REVIEW

File: rmw_fastrtps_cpp/src/rmw_init.cpp [hunk 1]
```diff
@@ -12,11 +12,47 @@
...
+  {
+  }
+
+  virtual ~RcutilsLogConsumer()
+  {
+  }
+
+  virtual void Consume(const Log::Entry & entry)
+  {
+    const char * logger_name = "rmw_fastrtps";
```
Comment (jubeira) [inline rmw_fastrtps_cpp/src/rmw_init.cpp]: This could be a static `constexpr`.

File: rmw_fastrtps_cpp/src/rmw_init.cpp [hunk 2]
```diff
@@ -77,6 +113,14 @@ rmw_init(const rmw_init_options_t * options, rmw_context_t * context)
     options->implementation_identifier,
     eprosima_fastrtps_identifier,
     return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
+  // Forward fastrtps log messages into rcl logging system.
+  LogConsumer * my_consumer = new RcutilsLogConsumer();
```
Comment (jubeira) [inline rmw_fastrtps_cpp/src/rmw_init.cpp]: Create the consumer directly inside the `unique_ptr` if that's the only use it has.

File: rmw_fastrtps_cpp/src/rmw_init.cpp [hunk 3]
```diff
@@ -77,6 +113,14 @@ rmw_init(const rmw_init_options_t * options, rmw_context_t * context)
...
+  Log::Reset();
+  Log::ClearConsumers();
+  Log::RegisterConsumer(std::unique_ptr<LogConsumer>(my_consumer));
+  Log::SetVerbosity(Log::Info);
```
Comment (jubeira) [inline rmw_fastrtps_cpp/src/rmw_init.cpp]: This should be configurable at some point.

File: rmw_fastrtps_cpp/src/rmw_init.cpp [hunk 4]
```diff
@@ -12,11 +12,47 @@
...
-
 #include "rmw_fastrtps_cpp/identifier.hpp"
 
+using eprosima::fastrtps::LogConsumer;
+using eprosima::fastrtps::Log;
+
+class RcutilsLogConsumer : public LogConsumer
+{
+public:
+  RcutilsLogConsumer()
```
Comment (jubeira) [inline rmw_fastrtps_cpp/src/rmw_init.cpp]: I'd change to `= default`.

**Comment by jubeira** [comment]
Hi @windelbouwman, did you have the chance to check my previous comment?

> Do you have a use-case and a simple way to test this feature? I've compiled FastRTPS in Debug mode and enabled debug messages in its CMakeLists file, but I still couldn't see any messages coming out when creating nodes and publishing to topics. Am I missing something?

Thanks!

**Comment by mabelzhang** [comment]
@windelbouwman, did you still want to continue on this PR?

**Comment by windelbouwman** [comment]
@mabelzhang I'm unsure. I still do not know what is the best strategy to forward the fastrtps logging. I would like to include it into the ros logging, so all logging is bundled in one place. Hence this PR. Feel free to close the pull request. I wanted to ask the question whether this was done already or not.

**Comment by gonzodepedro** [comment]
This is not done already. Feel free to reopen with addressed comments and a use-case.

**Pull request closed without merging**